src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,297 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json.Serialization.Converters;
+
+namespace System.Text.Json.Node
+{
+    /// <summary>
+    ///   Represents a mutable JSON object.
+    /// </summary>
+    [DebuggerDisplay("JsonObject[{Dictionary.Count}]")]
+    [DebuggerTypeProxy(typeof(DebugView))]
+    public sealed partial class JsonObject : JsonNode
+    {
+        private JsonElement? _jsonElement;
+        private IDictionary<string, JsonNode?>? _dictionary;
+        private string? _lastKey;
+        private JsonNode? _lastValue;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class that is empty.
+        /// </summary>
+        /// <param name="options">Options to control the behavior.</param>
+        public JsonObject(JsonNodeOptions? options = null) : base(options) { }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonObject"/> class that contains the specified <paramref name="properties"/>.
+        /// </summary>
+        /// <param name="properties">The properties to be added.</param>
+        /// <param name="options">Options to control the behavior.</param>
+        public JsonObject(IEnumerable<KeyValuePair<string, JsonNode?>> properties, JsonNodeOptions? options = null)
+        {
+            foreach (KeyValuePair<string, JsonNode?> node in properties)
+            {
+                Add(node.Key, node.Value);
```

Comment:
Turning `_dictionary` into a cocrete type means that we could probably use the `Dictionary<K,V>.AddRange` method here, which performs a few optimizations depending on the input.

Reply:
Sure. When this PR is in, I'll switch over to a custom type anyway (for deterministic ordering semantics) that internally will use a dictionary with direct access.